File: ios/WorkSphere/WorkSphere/Services/AuthService.swift
import Foundation
import SwiftUI

@Observable
final class AuthService {
    static let shared = AuthService()

    var isAuthenticated = false
    var userId: Int64?
    var username: String?
    var isAdmin = false
    var debugMode = false

    private let api = APIClient.shared

    init() {
        // Restore from UserDefaults
        if let token = UserDefaults.standard.string(forKey: "auth_token") {
            api.token = token
            let stored = UserDefaults.standard.object(forKey: "auth_userId")
            if let num = stored as? Int64 { userId = num }
            else if let num = stored as? Int { userId = Int64(num) }
            else if let str = stored as? String { userId = Int64(str) }
            username = UserDefaults.standard.string(forKey: "auth_username")
            isAdmin = UserDefaults.standard.bool(forKey: "auth_isAdmin")
            if userId != nil { isAuthenticated = true }
        }
        let debugStored = UserDefaults.standard.object(forKey: "debug_userId")
        var debugId: Int64?
        if let num = debugStored as? Int64 { debugId = num }
        else if let num = debugStored as? Int, num > 0 { debugId = Int64(num) }
        else if let str = debugStored as? String { debugId = Int64(str) }
        if let debugId, debugId > 0 {
            api.debugUserId = debugId
            userId = debugId
            debugMode = true
            isAuthenticated = true
        }
    }

    func login(username: String, password: String) async throws {
        let response: LoginResponse = try await api.post("/auth/login", body: LoginRequest(username: username, password: password))
        let uid = response.userId
        await MainActor.run {
            api.token = response.token
            api.debugUserId = nil
            self.userId = uid
            self.username = response.username
            self.isAdmin = response.admin
            self.isAuthenticated = true
            self.debugMode = false
            persist(token: response.token, userId: uid, username: response.username, isAdmin: response.admin)
        }
    }

    func register(username: String, displayName: String, email: String, password: String, bio: String?) async throws {
        let response: LoginResponse = try await api.post("/auth/register", body: RegisterRequest(
            username: username, displayName: displayName, email: email, password: password, bio: bio
        ))
        let uid = response.userId
        await MainActor.run {
            api.token = response.token
            api.debugUserId = nil
            self.userId = uid
            self.username = response.username
            self.isAdmin = response.admin
            self.isAuthenticated = true
            self.debugMode = false
            persist(token: response.token, userId: uid, username: response.username, isAdmin: response.admin)
        }
    }

    func loginDebug(userId: Int64) {
        api.token = nil
        api.debugUserId = userId
        self.userId = userId
        self.debugMode = true
        self.isAuthenticated = true
        UserDefaults.standard.set(userId, forKey: "debug_userId")
    }

    func logout() {
        api.token = nil
        api.debugUserId = nil
        userId = nil
        username = nil
        isAdmin = false
        isAuthenticated = false
        debugMode = false
        UserDefaults.standard.removeObject(forKey: "auth_token")
        UserDefaults.standard.removeObject(forKey: "auth_userId")
        UserDefaults.standard.removeObject(forKey: "auth_username")
        UserDefaults.standard.removeObject(forKey: "auth_isAdmin")
        UserDefaults.standard.removeObject(forKey: "debug_userId")
    }

    func configureServer(url: String) {
        api.baseURL = url + "/api"
    }

    private func persist(token: String, userId: Int64, username: String, isAdmin: Bool) {
        UserDefaults.standard.set(token, forKey: "auth_token")
        UserDefaults.standard.set(String(userId), forKey: "auth_userId")
        UserDefaults.standard.set(username, forKey: "auth_username")
        UserDefaults.standard.set(isAdmin, forKey: "auth_isAdmin")
    }
}

File: android/WorkSphere/app/src/main/java/com/worksphere/app/data/AuthService.kt
package com.worksphere.app.data

import android.content.Context
import android.content.SharedPreferences
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/**
 * Singleton authentication manager.
 *
 * Call [init] once from Application/Activity before using any other method.
 * Credentials are persisted in SharedPreferences so the user stays logged in
 * across app restarts.
 */
object AuthService {

    private const val PREFS_NAME = "worksphere_auth"
    private const val KEY_TOKEN = "token"
    private const val KEY_USER_ID = "user_id"
    private const val KEY_USERNAME = "username"
    private const val KEY_IS_ADMIN = "is_admin"
    private const val KEY_DEBUG_MODE = "debug_mode"
    private const val KEY_DEBUG_USER_ID = "debug_user_id"
    private const val KEY_SERVER_URL = "server_url"
    private const val KEY_TENANT_ID = "tenant_id"

    private lateinit var prefs: SharedPreferences

    private val _isAuthenticated = MutableStateFlow(false)
    val isAuthenticated: StateFlow<Boolean> = _isAuthenticated.asStateFlow()

    var token: String? = null
        private set
    var userId: Long = 0L
        private set
    var username: String = ""
        private set
    var isAdmin: Boolean = false
        private set
    var debugMode: Boolean = false
        private set
    var tenantId: String? = null
        private set

    // ---------------------------------------------------------------------
    // Initialisation
    // ---------------------------------------------------------------------

    fun init(context: Context) {
        prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        restoreSession()
    }

    private fun restoreSession() {
        val savedUrl = prefs.getString(KEY_SERVER_URL, null)
        if (savedUrl != null) {
            ApiClient.baseUrl = savedUrl
        }

        // Restore tenant
        val savedTenantId = prefs.getString(KEY_TENANT_ID, null)
        if (savedTenantId != null) {
            tenantId = savedTenantId
            ApiClient.tenantId = savedTenantId
        }

        val savedToken = prefs.getString(KEY_TOKEN, null)
        val savedDebugMode = prefs.getBoolean(KEY_DEBUG_MODE, false)
        val savedUserId = prefs.getLong(KEY_USER_ID, 0L)

        if (savedDebugMode && savedUserId != 0L) {
            debugMode = true
            userId = savedUserId
            username = prefs.getString(KEY_USERNAME, "") ?: ""
            isAdmin = prefs.getBoolean(KEY_IS_ADMIN, false)
            ApiClient.debugUserId = savedUserId
            ApiClient.token = null
            _isAuthenticated.value = true
        } else if (savedToken != null) {
            token = savedToken
            userId = savedUserId
            username = prefs.getString(KEY_USERNAME, "") ?: ""
            isAdmin = prefs.getBoolean(KEY_IS_ADMIN, false)
            debugMode = false
            ApiClient.token = savedToken
            ApiClient.debugUserId = null
            _isAuthenticated.value = true
        }
    }

    // ---------------------------------------------------------------------
    // Login / Register / Logout
    // ---------------------------------------------------------------------

    /**
     * Authenticate with username + password. Throws on failure.
     */
    suspend fun login(username: String, password: String): LoginResponse {
        val response = ApiClient.post<LoginResponse>(
            path = "/auth/login",
            body = LoginRequest(username, password)
        )

        applyLoginResponse(response)
        return response
    }

    /**
     * Register a new account and immediately log in.
     */
    suspend fun register(
        username: String,
        displayName: String,
        email: String,
        password: String,
        bio: String? = null
    ): LoginResponse {
        val response = ApiClient.post<LoginResponse>(
            path = "/auth/register",
            body = RegisterRequest(username, displayName, email, password, bio)
        )

        applyLoginResponse(response)
        return response
    }

    /**
     * Debug / impersonation login – skips real auth, sets X-Debug-User-Id header.
     */
    suspend fun loginDebug(debugUserId: Long) {
        // Validate the user exists by fetching their profile
        val user = ApiClient.get<UserDto>(
            path = "/users/$debugUserId",
            queryParams = emptyMap()
        )

        token = null
        userId = user.id
        username = user.username
        isAdmin = user.admin ?: false
        debugMode = true

        ApiClient.token = null
        ApiClient.debugUserId = debugUserId

        prefs.edit()
            .putString(KEY_TOKEN, null)
            .putLong(KEY_USER_ID, userId)
            .putString(KEY_USERNAME, username)
            .putBoolean(KEY_IS_ADMIN, isAdmin)
            .putBoolean(KEY_DEBUG_MODE, true)
            .putLong(KEY_DEBUG_USER_ID, debugUserId)
            .putString(KEY_TENANT_ID, tenantId)
            .apply()

        _isAuthenticated.value = true
    }

    fun logout() {
        token = null
        userId = 0L
        username = ""
        isAdmin = false
        debugMode = false
        tenantId = null

        ApiClient.token = null
        ApiClient.debugUserId = null
        ApiClient.tenantId = null

        prefs.edit().clear().apply()
        _isAuthenticated.value = false
    }

    // ---------------------------------------------------------------------
    // Tenant selection
    // ---------------------------------------------------------------------

    fun setTenant(id: String?) {
        tenantId = id
        ApiClient.tenantId = id
    }

    // ---------------------------------------------------------------------
    // Server configuration
    // ---------------------------------------------------------------------

    fun configureServer(url: String) {
        val trimmed = url.trimEnd('/')
        ApiClient.baseUrl = trimmed
        prefs.edit().putString(KEY_SERVER_URL, trimmed).apply()
    }

    fun getServerUrl(): String = ApiClient.baseUrl

    // ---------------------------------------------------------------------
    // Internal
    // ---------------------------------------------------------------------

    private fun applyLoginResponse(response: LoginResponse) {
        token = response.token
        userId = response.userId
        username = response.username
        isAdmin = response.admin
        debugMode = false

        ApiClient.token = response.token
        ApiClient.debugUserId = null

        prefs.edit()
            .putString(KEY_TOKEN, response.token)
            .putLong(KEY_USER_ID, response.userId)
            .putString(KEY_USERNAME, response.username)
            .putBoolean(KEY_IS_ADMIN, response.admin)
            .putBoolean(KEY_DEBUG_MODE, false)
            .putString(KEY_TENANT_ID, tenantId)
            .apply()

        _isAuthenticated.value = true
    }
}
